internal/ingress/controller/controller.go

```diff
@@ -200,6 +200,11 @@ func (n *NGINXController) syncIngress(interface{}) error {
 		}(isFirstSync)
 	}
 
+	re := getRemovedEndpoints(n.runningConfig, &pcfg)
+	if len(re) > 0 {
+		go n.metricCollector.RemoveMetrics(re)
```

Comment:
It’s important to have a test that ensures this operation is atomic. Because the goroutine is left unmonitored it could _potentially_ run concurrently (eg. the sync fails and gets retried while metrics are still being removed).

Reply:
> sync fails and gets retried while metrics are still being removed

At this point, the sync already happened.
Also, syncIngress is not executed concurrently (and also is rate limited)